plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+module Language.Plutus.Contract.Contract where
+
+import           Control.Applicative (liftA2)
+import           Control.Monad       ((>=>))
+import           Data.Bifunctor      (first)
+import           Data.List           (foldl')
+
+data Contract t i a =
+    Waiting (i -> Contract t i a)
+    | Done
+    | Emit t (Contract t i a) -- produce a 't' value
+    | Pure a
+    deriving (Functor)
+
+-- The applicative instance parallelises the 'Waiting' operations
+instance Applicative (Contract t i) where
+    pure = Pure
+    cf <*> ca = case cf of
+        Done -> Done
+        Emit t c -> Emit t (c <*> ca)
+        Pure f -> case ca of
```

Comment:
Isn't this case just `fmap f ca`? Might need to write a separate `Functor` instance.

Reply:
I think it's just `fmap f ca`.